### ISSUE TITLE
Can't scan folder folder on a mounted NFS share

### ISSUE THREAD

**Issue by terslowe** [photoview/photoview#1104]
I get this error on the webgui

`Something went wrong
Server error: failed to add user for scanning (1): find albums for user (user_id: 1): read directory (/photos): open /photos: permission denied at (scanAll)`

1. i have mounted a nfs share on the linux machine where the docker is installed.

'mount -t nfs -O "tcp,actimeo=1800,gid=1001,uid=1001" 192.168.100.5:/NFS /mnt/nfs' 

2. I can browse via de linux machine to the nfs share without problem and list and read the files.

3. i have checked the .env file and put there the /mnt/nfs and all config seems correct.

I think the problem should be that docker or the application access problem with uid/gid option.

**Comment by kkovaletp** [comment]
@terslowe, as far as I understand, the UID and GID in the mount command are for the remote server. Once the folder is mounted, its permissions and content are transparent for apps and services on the system, including Docker.
Most probably, the solution is described in the https://github.com/photoview/photoview?tab=readme-ov-file#getting-started--setup-with-docker point 4.

**Comment by terslowe** [comment]
Okay, i have checked that and make some tests.

Why do i really need 999:999 UID and GID? Can i use another UID and GID and give the docker compose to startup with anoter user maybe? like in de yml file put in "user:1001:1001"

i can't use 999 on my NAS because its already used on the Nas default by another default user.

**Comment by kkovaletp** [comment]
> Why do i really need 999:999 UID and GID?

Because this is how our image is built, it won't work with other IDs if you set them in the compose file.
Feel free to propose a PR with the fix to allow user-defined IDs from the compose file. I guess that this won't be an easy task...

> i can't use 999 on my NAS because its already used on the Nas default by another default user.

You don't have to. You can just set read permissions to the `others` group for the mounted folder and all its content recursively. It should be enough for Photoview to access and read your media.

**Comment by terslowe** [comment]
And maybe because of this issue takes me to long what are the better secure options to share and mount network folder to photoview?

**Comment by kkovaletp** [comment]
> what are the better secure options to share and mount network folder

Well, if you prefer NFS, use the latest version of NFS 4 with Kerberos authentication. However, it is a bit tricky to configure, so I'd recommend using Samba, as it supports secure-auth out-of-the-box. If you want even more security, you can set up a cloud storage server with MFA and strong HTTPS encryption, but you'd also need to install a client on the Photoview host—Linux `mount` doesn't support such protocols natively.

As I mentioned earlier, once the network share/export is mounted on the host, it behaves as a regular folder for all other apps. A local file system on the host is usually considered to be a safe place, so granting read permissions to the `others` group shouldn't be a big issue. However, if you'd like to control the access to it locally, you can mount this network share/export to a sub-folder of a folder, where only Photoview can access - in this case even if you grant the `777` permissions to the mounted folder recursively, nobody else can even see it

**Comment by terslowe** [comment]
> > Why do i really need 999:999 UID and GID?
> 
> Because this is how our image is built, it won't work with other IDs if you set them in the compose file. Feel free to propose a PR with the fix to allow user-defined IDs from the compose file. I guess that this won't be an easy task...
> 
> > i can't use 999 on my NAS because its already used on the Nas default by another default user.
> 
> You don't have to. You can just set read permissions to the `others` group for the mounted folder and all its content recursively. It should be enough for Photoview to access and read your media.

1. i did give others read rights but can't scan the photos inside photoview
2. Read Write to the others also not helps.
3. Read Write Execute to the others i can scan the folders.

There is i think some problems what i can not find it with the permissions on Photoview.

**Comment by terslowe** [comment]
I have also tried immich with the same NFS permissions that is given from the server, and on this app i can well scan the folder. This likes me really a problem with photoview app. Maybe this helps to find the problem?

**Comment by kkovaletp** [comment]
> * i did give others read rights but can't scan the photos inside photoview
> * Read Write to the others also not helps.
> * Read Write Execute to the others i can scan the folders.

I'm sorry to be not crystal clear, assuming you know how Linux FS permissions work.

Even while both files and folders have the same RWX permission types for the same 3 groups (owner, group, others), they have a bit different impact on files and folders. In particular, to read all the content of a file, it is enough to give the R permission for the corresponding group. However, to be able to read all the content of a folder (including the ability to navigate inside and read the content of all sub-items), you need to give both R and X permissions. That is why in many UIs and documentation it has the full name 'Execute and Search'.

So, basically, for our case you need to run 2 commands, listed in the Makefile example in lines 79 and 80 (or run `make readable` if you use the Makefile approach): https://github.com/photoview/photoview/blob/84c642ca79b3a1f2f51cc3c2d8318818f202acfc/docker-compose%20example/Makefile#L79
Of course, you might need to replace the env variable with the full path to the media library root folder.

**Comment by kkovaletp** [comment]
> I have also tried immich with the same NFS permissions that is given from the server, and on this app i can well scan the folder. This likes me really a problem with photoview app. Maybe this helps to find the problem?

I know that other docker containers are able to access the host FS with no additional modifications of host permissions. This is because many of them run under the `root` user, which is a huge security problem. We build our image to run under a non-root user, which doesn't have full access to the FS respectively to the permissions. That is why we have to set the correct permissions on the host.

This is not a problem. This is a feature, and this is the correct way, an app should run in a Linux environment. As I mentioned earlier, in theory, there should be a way to configure UID and GID for the Photoview user inside the container, so that users are able to align these IDs between the host and the container. However, in practice, I don't see an easy way to implement this. If someone would propose a good solution for this feature, we'd be happy to accept and merge it ;-)

**Comment by terslowe** [comment]
Can you please describe the comment below how to do that?

`if you'd like to control the access to it locally, you can mount this network share/export to a sub-folder of a folder, where only Photoview can access`

**Comment by kkovaletp** [comment]
Sure. Here is 1 of the possible implementations of this idea:
- Assuming that you have the `/media` folder on your host with `drwxr-xr-x` permissions, where you want the source media lib to be mounted
- Create a local user on the host with UID:GID = 999:999 - same as the `photoview` user inside the container.
- Instead of creating the mount point right here (where everyone can see it), let's create the `/media/Photoview` folder, owned by the local user from the previous step. Then set the `drwx------` permissions to it, so it is accessible only by this local user.
- Finally, create the mountpoint folder inside this `/media/Photoview` folder (for example, `media`) and set any permissions to it and its content, you'd like, even `rwxrwxrwx` - only the local user with UID:GID = 999:999 will be able to access them on the host because other users (except for the `root` user, of course) won't be able to dive into the parent `Photoview` folder.

**Comment by kkovaletp** [comment]
Feel free to continue our discussion here if you need some more info/suggestions.
But as it is clear at this point that this is not an issue with the product, but a misconfiguration case, I'm closing it as not planned.
Again, we still can continue the communication, even when the issue is closed.
